### DISCUSSION TITLE
RPC functions needs to know the sender's node id

### DISCUSSION THREAD

**Discussion by songweijia** [Derecho-Project/derecho#227] [Ideas]
Sometimes, the RPC function needs to know who is the caller. In DDS, for example, the server needs to know the id of an external client who trying to register itself to some topic, so that the server can send a notification to it. However, RPC functions are unaware of who is the sender.

I plan to enable this feature by adding a thread-local static member in RPCManager to keep the caller id for both p2p calls and ordered calls. And this thread-local member will be exposed to the RPC functions through `GroupReference::group`: a mechanism similar to `_Group::get_num_of_shards()`. I will implement this feature on a fork from the `client_callbacks` branch. Please let me know what do you think.

**Comment by songweijia** [comment]
Actually, it can be just a static function `RPCManager::get_rpc_caller_id()`.

**Comment by KenBirman** [comment]
I think this would be a valuable addition to the system

**Reply by songweijia** [reply]
I created a pull request implementing this feature. @etremel Edward, I worked on your branch, so it would be better for you to review and merge it.

**Comment by etremel** [comment]
I agree that RPC functions should have a way of determining the ID of the node that called them. It would be nice if it could somehow be passed in to the RPC function as an argument, since at the point RPCManager invokes the RPC function it already knows the caller's ID (and `receive_call` already has a mandatory "caller ID" parameter), but that would probably take a lot of effort to implement because we have no control over the signature of RPC-callable methods. Providing a function you can call within the RPC method is probably the next-best option.

**Reply by songweijia** [reply]
Oh, I have exposed the caller ID already in the pull request I created here: https://github.com/Derecho-Project/derecho/pull/228/files, which happens to be the next-best option you mentioned.

**Reply by songweijia** [reply]
Basically, I created a thread_local static member for `RPCManager` class and provide a static method `RPCManager::get_rpc_caller_id()` to read it. Inside an RPC function, calling this is always valid because the scope is inside the thread which set the thread-local static `RPCManager::caller_id` immediately before calling the RPC function.
